### ISSUE TITLE
Support for pattern-matched rules

### ISSUE THREAD

**Issue by solnic** [dry-rb/dry-validation#678]
In order to make defining rules that depend on additional conditions nicer, we can have something like this:

```ruby
params do
  optional(:per_page).maybe(:integer)
end

rule(:per_page).if(:key?).validate(gt?: 0, lteq?: 20)
```

See the original conversation: https://github.com/dry-rb/dry-validation/issues/540#issuecomment-744022950

**Comment by musaffa** [comment]
Above code won't work. For `optional`, we will need `:key?`. What is going to check for `maybe`? Unlike `optional` which is related to `key`, `maybe` is related to its `value`.

**Comment by solnic** [comment]
@musaffa right, thanks - just updated the example. The actual part that needs to be implemented is the `if` method. The underlying predicates (like `key?`) will be available OOTB, so we can add `value?` like you suggested in a separate PR, because essentially it's going to be a separate, new feature as well 🙂

**Comment by musaffa** [comment]
@solnic Hi, Is there any update about this feature?

**Comment by solnic** [comment]
@musaffa hey there, no, nobody has had the time to work on this yet. There's a lot of more important things to do in validation/schema so this needs to wait.

**Comment by musaffa** [comment]
@solnic I can help with implementing this feature. For that, I'll need some guidance from you.

**Comment by solnic** [comment]
@musaffa that's cool! Feel free to grab this issue. I don't really have any specific idea how to implement this so you need to experiment a bit. Rules are captured via [`Contract.rule`](https://github.com/dry-rb/dry-validation/blob/0efa0b90620d3bd11e0eed4d2d3146f634f8f862/lib/dry/validation/contract/class_interface.rb#L107) so I think something similar can be done as we have in case of [`Rule#each`](https://github.com/dry-rb/dry-validation/blob/0efa0b90620d3bd11e0eed4d2d3146f634f8f862/lib/dry/validation/rule.rb#L79-L96). The trick is that `each` overrides rule's block, and with this `if` method it should most likely create a rule-like object that would just serve as a guard. Maybe something like this:

```ruby
# inside rule.rb
def if(*args)
  Guard.new(self, args: args)
end
```

This `Guard` thingie would have to quack like a rule, using a decorator should work better than inheriting from Rule. You also need to figure out how to *replace* the rule with the guard because `Contract.rule` adds rule instances to the internal `rules` array. This can be a bit difficult though so...another option is to use a different top-level DSL method for this, like `guard(:key?).rule(:foo).validate(:good?)`.

Yeah...this needs some experimentation to see what makes more sense 🙂

**Comment by musaffa** [comment]
Thanks. I'll look into it.

**Comment by solnic** [comment]
@musaffa I'd start with this additional top-level method, actually 🙂

**Comment by musaffa** [comment]
Noted.

**Comment by ssoulless** [comment]
Any news on this? would this also be covering rules that has more than one key?

```
rule(:key1, :key2) do
    ...
 end
```
